### PR TITLE
feat: add scopes

### PR DESCRIPTION
**Pull request by oney**
Add scopes feature like ORMs in Ruby on Rails and Laravel.
See document https://github.com/oney/typeorm/blob/add-scopes/docs/scope.md

Please give any feedback or idea.
Might need more tests.

### CODE REVIEW

File: src/decorator/scopes/Scope.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import {getMetadataArgsStorage} from "../../";
+import {ScopeMetadataArgs} from "../../metadata-args/ScopeMetadataArgs";
+
+export function Scope(global: boolean = false): Function {
```
Comment (Ginden) [inline src/decorator/scopes/Scope.ts:4]: `MethodDecorator`

File: src/find-options/FindOneOptions.ts [hunk 2]
```diff
@@ -32,6 +33,10 @@ export interface FindOneOptions<Entity = any> {
      */
     order?: { [P in keyof Entity]?: "ASC"|"DESC"|1|-1 };
 
+    unscoped?: boolean;
```
Comment (Ginden) [inline src/find-options/FindOneOptions.ts:36]: I would prefer `scoped` with default `true`.

File: src/decorator/scopes/Scope.ts (hunk 1 above)
Comment (Ginden) [inline src/decorator/scopes/Scope.ts:4]: Boolean parameters are usually considered bad. `{global: boolean}` will be future-proof.

File: src/metadata/ScopeMetadata.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    // ---------------------------------------------------------------------
+    // Public Build Methods
+    // ---------------------------------------------------------------------
+
+    /**
+     * Builds some depend scope properties.
+     */
+    build(): this {
+        return this;
```
Comment (Ginden) [inline src/metadata/ScopeMetadata.ts:56]: Why is it needed?

File: src/query-builder/SelectQueryBuilder.ts [hunk 4]
```diff
@@ -2027,4 +2031,32 @@ export class SelectQueryBuilder<Entity> extends QueryBuilder<Entity> implements
...
+        this.expressionMap.globalScoped = globalScoped;
+        return this;
+    }
+
+    scope(scopeFn: ScopeFn<this>): this;
+    scope(scopeFns: Array<ScopeFn<this>>): this;
+    scope(scopeFn:  ScopeFn<this> | Array<ScopeFn<this>>): this {
+        const scopeFns = Array.isArray(scopeFn) ? scopeFn : [scopeFn];
+        scopeFns.forEach(scopeFn => {
+            scopeFn(this);
```
Comment (Ginden) [inline src/query-builder/SelectQueryBuilder.ts:2058]: Should provide alias as second parameter - two entities can have the same field name.

**Comment by pleerock** [comment]
According to author last comments in a different threads it doesn't seem he is interesting in doing anything in this repository.

**Pull request closed without merging**